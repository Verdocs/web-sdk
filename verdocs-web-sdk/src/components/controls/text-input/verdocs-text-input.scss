verdocs-text-input {
  /* Fallback font for old browsers */
  font-family: sans-serif;
  font-family: var(--verdocs-font);
  box-sizing: border-box;

  .input-field {
    box-sizing: border-box;
    margin: 20px 0;
  }

  .input-element {
    box-sizing: border-box;
    border: 1px solid #707ae5;
    border-radius: 4px;
    background: #fff;
    color: #000;
    width: 100%;
    font-size: 16px;
    height: 28px;
    padding: 4px 12px;

    &::placeholder {
      color: #aaa;
      opacity: 1;
    }
  }

  .input-label {
    display: block;
    color: #333;
    font-weight: 400;
    //color: #aeb4bf;
    font-size: 14px;
    margin: 0 0 7px 0;
  }
}
