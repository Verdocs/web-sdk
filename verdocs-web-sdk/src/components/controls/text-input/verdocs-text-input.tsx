import {Component, Prop, State, h, Event, EventEmitter} from '@stencil/core';

/**
 * Display a text input field. This is just a standard HTML input field with minimal markup to fit the
 * visual styles of the other components. Bear in mind Stencil does not currently support emitting events
 * with the same names as native DOM events, so rather than `input` or `blur` we have `tinput`, `tblur`,
 * etc. We used a temporary prefix here to mark this as a temporary API, because we hope this will change
 * in a future Stencil release. See https://github.com/ionic-team/stencil/issues/2822.
 */
@Component({
  tag: 'verdocs-text-input',
  styleUrl: 'verdocs-text-input.scss',
})
export class VerdocsTextInput {
  /**
   * The value for the input field.
   */
  @Prop() value: string = '';

  /**
   * The label for the field.
   */
  @Prop() label: string = '';

  /**
   * The placeholder for the field.
   */
  @Prop() placeholder: string = '';

  /**
   * If desired, the autocomplete attribute to set.
   */
  @Prop() autocomplete: string = '';

  /**
   * The type of field to render. Only text-type fields are allowed here for the current styling. Additional types
   * (e.g. a date picker) will be supported by other controls in the future.
   */
  @Prop() type: 'text' | 'password' | 'email' | 'number' | 'search' | 'tel' | 'url' = 'text';

  @State() _value: string = '';

  /**
   * Event fired when the input value changes.
   */
  @Event({composed: true}) tinput: EventEmitter<string>;

  /**
   * Event fired when the field receives focus.
   */
  @Event({composed: true}) tfocus: EventEmitter;

  /**
   * Event fired when the field loses focus.
   */
  @Event({composed: true}) tblur: EventEmitter;

  componentWillLoad() {
    this._value = this.value;
  }

  componentDidLoad() {}

  handleInput(event) {
    this._value = event.target.value;
    this.tinput.emit(this._value);
  }

  handleFocus(event) {
    this.tfocus.emit(event);
  }

  handleBlur(event) {
    this.tblur.emit(event);
  }

  render() {
    return (
      <div class="input-field">
        <label>
          {this.label ? <div class="input-label">{this.label + ':'}</div> : <div />}
          <input
            type={this.type}
            placeholder={this.placeholder}
            autoComplete={this.autocomplete}
            value={this._value}
            class="input-element"
            onInput={event => this.handleInput(event)}
            onFocus={event => this.handleFocus(event)}
            onBlur={event => this.handleBlur(event)}
          />
        </label>
      </div>
    );
  }
}
