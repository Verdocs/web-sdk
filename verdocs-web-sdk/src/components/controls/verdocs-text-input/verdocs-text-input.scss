@import '../../../theme.scss';

verdocs-text-input {
  font-family: $primary-font;
  box-sizing: border-box;
  margin: 20px 0;
  display: block;

  .input-element {
    box-sizing: border-box;
    border: 1px solid $border-color;
    border-radius: 4px;
    background: #fff;
    color: #000;
    width: 100%;
    font-size: 16px;
    height: 28px;
    padding: 4px 12px;

    &::placeholder {
      color: $light-disabled-color;
      opacity: 1;
    }
  }

  .input-label {
    display: block;
    color: $medium-bg;
    font-weight: 400;
    font-size: 14px;
    margin: 0 0 4px 0;
  }
}
