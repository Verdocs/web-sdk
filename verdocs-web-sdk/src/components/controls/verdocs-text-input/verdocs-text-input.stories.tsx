import {html} from 'lit-html';
import {Meta} from '@storybook/web-components';

export default {
  title: 'Controls/Text Input',
  component: 'verdocs-text-input',
  args: {
    label: 'Name',
    value: '',
    placeholder: 'Enter your name...',
  },
  argTypes: {
    input: {
      action: 'input',
      table: {
        disable: true,
      },
    },
  },
} as Meta;

export const TextInput = ({type, label, value, placeholder, disabled, input}) =>
  html`<verdocs-text-input .type=${type} .label=${label} .placeholder=${placeholder} .value=${value} .disabled=${disabled} @input=${input} />`;
