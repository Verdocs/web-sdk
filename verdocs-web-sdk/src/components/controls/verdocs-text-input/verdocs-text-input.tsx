import {Component, Element, Prop, Host, h} from '@stencil/core';
import {VerdocsToast} from '../../../utils/Toast';

const ClearIcon = `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 122.881 122.88"><g><path fill-rule="evenodd" clip-rule="evenodd" d="M61.44,0c33.933,0,61.441,27.507,61.441,61.439 c0,33.933-27.508,61.44-61.441,61.44C27.508,122.88,0,95.372,0,61.439C0,27.507,27.508,0,61.44,0L61.44,0z M81.719,36.226 c1.363-1.363,3.572-1.363,4.936,0c1.363,1.363,1.363,3.573,0,4.936L66.375,61.439l20.279,20.278c1.363,1.363,1.363,3.573,0,4.937 c-1.363,1.362-3.572,1.362-4.936,0L61.44,66.376L41.162,86.654c-1.362,1.362-3.573,1.362-4.936,0c-1.363-1.363-1.363-3.573,0-4.937 l20.278-20.278L36.226,41.162c-1.363-1.363-1.363-3.573,0-4.936c1.363-1.363,3.573-1.363,4.936,0L61.44,56.504L81.719,36.226 L81.719,36.226z"/></g></svg>`;
const CopyIcon = `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20"><path fill-rule="evenodd" d="M15.988 3.012A2.25 2.25 0 0118 5.25v6.5A2.25 2.25 0 0115.75 14H13.5V7A2.5 2.5 0 0011 4.5H8.128a2.252 2.252 0 011.884-1.488A2.25 2.25 0 0112.25 1h1.5a2.25 2.25 0 012.238 2.012zM11.5 3.25a.75.75 0 01.75-.75h1.5a.75.75 0 01.75.75v.25h-3v-.25z" clip-rule="evenodd" /><path fill-rule="evenodd" d="M2 7a1 1 0 011-1h8a1 1 0 011 1v10a1 1 0 01-1 1H3a1 1 0 01-1-1V7zm2 3.25a.75.75 0 01.75-.75h4.5a.75.75 0 010 1.5h-4.5a.75.75 0 01-.75-.75zm0 3.5a.75.75 0 01.75-.75h4.5a.75.75 0 010 1.5h-4.5a.75.75 0 01-.75-.75z" clip-rule="evenodd" /></svg>`;

/**
 * Display a text input field. This is just a standard HTML input field with minimal markup to fit the
 * visual styles of the other components. Note that events "bubble" from the input field to the container,
 * so you can subscribe to the same DOM events (input, blur, etc) that a normal input would emmit.
 */
@Component({
  tag: 'verdocs-text-input',
  styleUrl: 'verdocs-text-input.scss',
})
export class VerdocsTextInput {
  private inputEl: HTMLInputElement;

  @Element()
  hostEl: HTMLElement;

  /**
   * The initial value for the input field.
   */
  @Prop({mutable: true, reflect: true}) value: string = '';

  /**
   * The label for the field.
   */
  @Prop() label: string = '';

  /**
   * The placeholder for the field.
   */
  @Prop() placeholder: string = '';

  /**
   * If desired, the autocomplete attribute to set.
   */
  @Prop() autocomplete: string = '';

  /**
   * If supplied, a help icon will be displayed to provide the user more information.
   */
  @Prop() helpText: string = '';

  /**
   * If set, a clear button will be displayed.
   */
  @Prop() clearable = false;

  /**
   * If set, a copy-to-clipboard button will be displayed. NOTE: A field may not be
   * both clearable and copyable. If both properties are set to true, copyable will
   * be ignored.
   */
  @Prop() copyable = false;

  /**
   * The type of field to render. Only text-type fields are allowed here for the current styling. Additional types
   * (e.g. a date picker) will be supported by other controls in the future.
   */
  @Prop() type: 'text' | 'password' | 'email' | 'number' | 'search' | 'tel' | 'url' = 'text';

  /**
   * Should the field be disabled?
   */
  @Prop() disabled: boolean = false;

  /**
   * Should the field be required?
   */
  @Prop() required: boolean = false;

  copyToClipboard() {
    this.inputEl.select();
    navigator.clipboard
      .writeText(this.value)
      .then(() => VerdocsToast('Copied!'))
      .catch(e => console.log('Error copying to clipboard', e));
  }

  render() {
    return (
      <Host class={`input-field ${this.clearable ? 'clearable' : ''} ${this.copyable ? 'copyable' : ''}`}>
        <label>
          {this.label && (
            <div class="input-label">
              {this.label + ':'}
              {this.required && <span class="required">*</span>}
            </div>
          )}

          <div class="input-container">
            <input
              type={this.type}
              value={this.value}
              required={this.required}
              class="input-element"
              data-lpignore="true"
              disabled={this.disabled}
              placeholder={this.placeholder}
              autoComplete={this.autocomplete}
              ref={el => (this.inputEl = el as HTMLInputElement)}
              onInput={(e: any) => (this.value = e.target.value)}
            />
            {this.clearable && this.value && (
              <span
                innerHTML={ClearIcon}
                class="clear"
                onClick={() => {
                  // We need to allow the onInput event to fire first in case
                  // the user was focused on this field before clicking clear.
                  setTimeout(() => {
                    this.value = '';
                    this.inputEl?.setAttribute('value', '');
                    // We need to allow the value to "settle"
                    setTimeout(() => {
                      this.inputEl?.dispatchEvent(new Event('focusout'));
                      this.inputEl?.blur();
                    }, 50);
                  }, 50);
                }}
              />
            )}

            {!this.clearable && this.copyable && this.value && <span innerHTML={CopyIcon} class="copy" onClick={() => this.copyToClipboard()} />}
          </div>

          {this.helpText && <verdocs-help-icon text={this.helpText} />}
        </label>
      </Host>
    );
  }
}
