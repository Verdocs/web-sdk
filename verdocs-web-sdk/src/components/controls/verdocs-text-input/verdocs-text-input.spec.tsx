import { newSpecPage } from '@stencil/core/testing';
import { VerdocsTextInput } from './verdocs-text-input';

describe('verdocs-text-input', () => {
  it('renders with default props', async () => {
    const page = await newSpecPage({
      components: [VerdocsTextInput],
      html: `<verdocs-text-input></verdocs-text-input>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('renders with label, value, and placeholder', async () => {
    const page = await newSpecPage({
      components: [VerdocsTextInput],
      html: `<verdocs-text-input label="Name" value="Alice" placeholder="Enter your name"></verdocs-text-input>`,
    });
    expect(page.root).toMatchSnapshot();
    const input = page.root.querySelector('input');
    expect(input?.value).toBe('Alice');
    expect(input?.placeholder).toBe('Enter your name');
  });

  it('renders with clearable and helpText', async () => {
    const page = await newSpecPage({
      components: [VerdocsTextInput],
      html: `<verdocs-text-input clearable="true" help-text="Help!"></verdocs-text-input>`,
    });
    expect(page.root).toMatchSnapshot();
    expect(page.rootInstance.clearable).toBe(true);
    expect(page.rootInstance.helpText).toBe('Help!');
  });
});
