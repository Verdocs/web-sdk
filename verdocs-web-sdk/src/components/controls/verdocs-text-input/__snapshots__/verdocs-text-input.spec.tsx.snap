// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verdocs-text-input renders with clearable and helpText 1`] = `
<verdocs-text-input class="clearable input-field" clearable="true" help-text="Help!" value="">
  <label>
    <div class="input-container">
      <input autocomplete="" class="input-element" data-lpignore="true" placeholder="" type="text" value="">
    </div>
    <verdocs-help-icon text="Help!"></verdocs-help-icon>
  </label>
</verdocs-text-input>
`;

exports[`verdocs-text-input renders with default props 1`] = `
<verdocs-text-input class="input-field" value="">
  <label>
    <div class="input-container">
      <input autocomplete="" class="input-element" data-lpignore="true" placeholder="" type="text" value="">
    </div>
  </label>
</verdocs-text-input>
`;

exports[`verdocs-text-input renders with label, value, and placeholder 1`] = `
<verdocs-text-input class="input-field" label="Name" placeholder="Enter your name" value="Alice">
  <label>
    <div class="input-label">
      Name:
    </div>
    <div class="input-container">
      <input autocomplete="" class="input-element" data-lpignore="true" placeholder="Enter your name" type="text" value="Alice">
    </div>
  </label>
</verdocs-text-input>
`;
